esi_leap/console/websocketproxy.py

```diff
@@ -0,0 +1,164 @@
+        )
+        url = urlparse.urlparse(console_info["console_info"]["url"])
+        connect_info.host = url.hostname
+        connect_info.port = url.port
+
+        return connect_info
+
+    def _close_connection(self, tsock, host, port):
+        """takes target socket and close the connection."""
+        try:
+            tsock.shutdown(socket.SHUT_RDWR)
+        except OSError:
+            pass
+        finally:
+            if tsock.fileno() != -1:
+                tsock.close()
+                LOG.debug(
+                    "%(host)s:%(port)s: "
+                    "Websocket client or target closed" % {"host": host, "port": port}
+                )
+
+    def new_websocket_client(self):
+        """Called after a new WebSocket connection has been established."""
+        # Reopen the eventlet hub to make sure we don't share an epoll
+        # fd with parent and/or siblings, which would be bad
+        from eventlet import hubs
+
+        hubs.use_hub()
+
+        token = (
+            urlparse.parse_qs(urlparse.urlparse(self.path).query)
+            .get("token", [""])
+            .pop()
+        )
+
+        try:
+            connect_info = self._get_connect_info(token)
+        except Exception:
+            LOG.debug(traceback.format_exc())
+            raise
+
+        host = connect_info.host
+        port = connect_info.port
+
+        # Connect to the target
+        LOG.debug("Connecting to: %(host)s:%(port)s" % {"host": host, "port": port})
+        tsock = self.socket(host, port, connect=True)
+
+        # Start proxying
+        try:
+            if CONF.serialconsoleproxy.timeout > 0:
+                conn_timeout = connect_info.expires - timeutils.utcnow_ts()
+                LOG.debug("%s seconds to terminate connection." % conn_timeout)
+                threading.Timer(
+                    conn_timeout, self._close_connection, [tsock, host, port]
+                ).start()
+            self.do_proxy(tsock)
+        except Exception:
+            LOG.debug(traceback.format_exc())
+            raise
+        finally:
+            self._close_connection(tsock, host, port)
+
+    def socket(self, *args, **kwargs):
+        return websockifyserver.WebSockifyServer.socket(*args, **kwargs)
+
+    def send_head(self):
+        # This code is copied from this example patch:
+        # https://bugs.python.org/issue32084#msg306545
+        path = self.translate_path(self.path)
+        if os.path.isdir(path):
+            parts = urlparse.urlsplit(self.path)
+            if not parts.path.endswith("/"):
+                # Browsers interpret "Location: //uri" as an absolute URI
+                # like "http://URI"
+                if self.path.startswith("//"):
+                    self.send_error(
+                        HTTPStatus.BAD_REQUEST, "URI must not start with //"
+                    )
+                    return None
+
+        return super(ProxyRequestHandler, self).send_head()
```

Comment:
The problem this is meant to solve apparently was fixed [in Python 3.11](https://github.com/python/cpython/issues/87389#issuecomment-1162440179).

Reply:
Ah - I'll take this out and test again!